olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,191 @@
+        model_config_copy = deepcopy(model_config)
+        if input_model_hash is None:
+            from huggingface_hub import repo_info
+
+            if model_config.has_hf_config():
+                hf_hub_model_commit_id = repo_info(model_config.get_hf_model_name()).sha
+        else:
+            if model_config_copy.config.get("model_path") is not None:
+                model_config_copy.config["model_path"] = None
+        return hash_dict(
+            {
+                "hf_hub_model_commit_id": hf_hub_model_commit_id,
+                "model_config": model_config_copy.to_json(),
+                "pass_search_point": pass_search_point,
+                "input_model_hash": input_model_hash,
+            }
+        )
+
+    def exist_in_cloud_cache(self, output_model_hash: str):
+        logger.info("Checking cloud cache for model hash: %s", output_model_hash)
+        return self.container_client.get_blob_client(output_model_hash).exists()
+
+    def get_model_config_by_hash_key(self, output_model_hash: str):
+        model_config_blob = f"{output_model_hash}/model_config.json"
+        model_config_path = Path(self.cache_dir) / output_model_hash
+        model_config_path.mkdir(parents=True, exist_ok=True)
+        model_config_path = model_config_path / "model_config.json"
+
+        blob_client = self.container_client.get_blob_client(model_config_blob)
+        logger.info("Downloading %s to %s", model_config_blob, model_config_path)
+
+        with open(model_config_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
+            download_file.write(download_stream.readall())
+
+        # model config
+        with open(model_config_path) as file:
+            model_config_dict = json.load(file)
+            return ModelConfig.from_json(model_config_dict)
+
+    def upload_model_to_cloud_cache(self, output_model_hash: str, output_model_config: ModelConfig):
+        logger.info("Uploading model to cloud cache.")
+        model_path = output_model_config.config.get("model_path")
+
+        if model_path is None:
+            logger.error("Model path is not found in the output model config. Upload failed.")
+            return
+
+        model_path = Path(model_path)
+        model_config_copy = deepcopy(output_model_config)
+        model_config_copy.config["model_path"] = None
+
+        # upload model config file
+        model_config_bytes = json.dumps(model_config_copy.to_json()).encode()
+        with io.BytesIO(model_config_bytes) as data:
+            self.container_client.upload_blob(f"{output_model_hash}/model_config.json", data=data, overwrite=False)
+
+        # upload model file
+        model_blob = str(Path(output_model_hash) / "model")
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            model_path_file = Path(temp_dir) / "model_path.json"
+            with open(model_path_file, "w") as file:
+                file.write({"model_path": model_path.name})
```

Comment:
Isn't `model_path` local file/dir path here? Why record a local file/dir path? If this is just the name of the file/folder, why not directly in the path computed on line 139 below? Why bother making a separate file merely to store the model name?

Reply:
No matter which format that the model file is, a file or folder, it can have different names, model file name or folder name. Olive doesn't have a standard to unify all output model names. So this file is only used to track model name (file name or folder name). When downloading the model from cloud cache, this name will be updated in the model config.

Reply:
As I understand it, the cloud is already mimicking the local file/folder path and names. So, I am not seeing why we need another map to track the same information. Can't we just ask the cloud itself to iterate a path and get the required information? 

In fact, the map isn't even complete because you are storing only the root folder path in the map and not all the files within the folder. In cloud_cache_helper.py:151 you call `upload_blob` but that file is not registered in the model path.

Reply:
As for a model path field in Model Config, if it is a folder, it is a folder name. If it is a file, it is a file name. So I record its folder name or file name here. If using blob API to determine its path, it needs some additional steps like checking if it is a file or folder, what is its name something like this.

Reply:
If it is a folder, is there a need to record each file within the folder? If yes, then that's not happening.

In my experience using remote caches, having these secondary records (ones that try to mimic cache content) always cause more trouble than being useful. Especially, given the cost of asking the cache itself what it has is very small. These records tend to go out of sync and issues from it are very hard to debug. The model file might say one thing but the cache something entirely different. Consider the fact that remote caches are primarily asynchronous in nature and don't implement an exclusive lock. - any number of paralleling running pipelines could potentially be updating the records here. There is absolutely no guarantee that two workflows might not try to update the same cache at the same time. You might end up with a scenario where half the files are from one workflow and other half from another.

In short, I would recommend not depending on the model path and using the remote API. 

Another improvement to consider - Upload the output of the workflow (all results including the model_path file into a temporary remote folder and then renaming it. If you use a unique identifier in the temporary path that should avoid the potential collision (and overwrites) from other running parallel workflows.

Reply:
If it is a folder, the model path is just the folder name, it doesn't need to record each file. Even we check the cache itself rather than recording the model path, it is still hard to address the paralleling running concern right? 

blob is a flat structure. There is no direct api to determine if it is a folder or file. There is also another case where we have an onnx model with external data, the model path should be the onnx model file path, while the model files could include onnx model file and external data file which is a folder. The iteration way to determine model path doesn't work here.

The model cache is based on Pass level but not workflow level. Do you mean output of pass?

Reply:
Let's run with the solution for now. We will revisit if we have specific scenarios where the model_path.json goes out of sync with cache.